File: sakuralearn-backend/src/main/java/com/sakuralearn/sakuralearn_backend/service/ProgressService.java
package com.sakuralearn.sakuralearn_backend.service;

import com.sakuralearn.sakuralearn_backend.dto.request.LessonBlockProgressRequest;
import com.sakuralearn.sakuralearn_backend.dto.response.LessonBlockProgressResponse;
import com.sakuralearn.sakuralearn_backend.dto.response.LessonProgressResponse;
import java.util.UUID;
import java.util.List;

public interface ProgressService {
    double getCourseProgress(UUID userId, UUID courseId);
    void updateLastAccessed(UUID userId, UUID lessonId);
    
    // New methods for Module 3
    LessonBlockProgressResponse updateBlockProgress(UUID userId, UUID blockId, LessonBlockProgressRequest request);
    List<LessonBlockProgressResponse> getLessonBlocksProgress(UUID userId, UUID lessonId);
    void completeLesson(UUID userId, UUID lessonId);
    List<UUID> getCompletedLessonIds(UUID userId, UUID courseId);
    void recalculateCourseProgressForAllUsers(UUID courseId);
}

File: sakuralearn-frontend/src/services/progressService.js
import API from './api'

export const progressService = {
  getCourseProgress: (courseId) => API.get(`/progress/course/${courseId}`),
  updateLastAccessed: (lessonId) => API.patch(`/progress/lessons/${lessonId}/access`),
  updateBlockProgress: (blockId, data) => API.patch(`/progress/blocks/${blockId}`, data),
  getLessonBlocksProgress: (lessonId) => API.get(`/progress/lessons/${lessonId}/blocks`),
  completeLesson: (lessonId) => API.post(`/progress/lessons/${lessonId}/complete`),
  getCompletedLessons: (courseId) => API.get(`/progress/course/${courseId}/completed-lessons`),
}
